clang/lib/3C/ConstraintBuilder.cpp

```diff
@@ -406,6 +448,19 @@ class ConstraintGenVisitor : public RecursiveASTVisitor<ConstraintGenVisitor> {
                                 TypeVarInfo &TVI)
       : Context(Context), Info(I), CB(Info, Context), TVInfo(TVI) {}
 
+  bool VisitTypedefDecl(TypedefDecl* TD) { 
+      CVarSet empty;
+      auto PSL = PersistentSourceLoc::mkPSL(TD, *Context);
+      bool shouldCheck = !PtrToStructDef::containsPtrToStructDef(TD);
+      // If we haven't seen this typedef before, initialize it's entry in the
+      // typedef map. If we have seen it before, and we need to preserve the
+      // constraints contained within it
+      if (!Info.seenTypedef(PSL)) {
```

Comment:
What happens if you see a different definition of the same typedef? I know there was an issue where both files `#include` the same header which declares the typedef. But what if they repeat the definition inline, e.g.,

a.c:
```
typedef int *ptr_int_t;
extern int bar(ptr_int_t x);
int foo(ptr_int_t p) {
  return bar(p);
}
```
b.c
```
typedef int *ptr_int_t;
int bar(ptr_int_t x) { return *x; }
```
And what if they don't interact, and one file treats the typdef pointer safely, but the other doesn't? I might expect in that case that you'd rewrite in one file but not the other.

Reply:
Seems to work like you described.

Reply:
I guess the question is: What should happen here? That is, suppose we change `bar`'s body to be 
```
int bar(ptr_int_t x) { x = (int *)5; return *x; }
```
In that case, we won't rewrite the typedef in `b.c` but you will rewrite it as if safe in `a.c` ?

Reply:
Yes, what currently happens is they get treated as two different typedefs. Given that that's how the C's type system treats them (IE: you can have two typedefs w/ the same name in different files and have them have different types)

Reply:
The problem here is that you are passing a variable whose type is the first typedef to a function that expects a parameter of the second typedef. C (and Checked C) will not catch this. Should we?

Reply:
Oh I misunderstood your example. In this case since they interact it resolves constraints between them. The current code for the example results in no conversions at all.

Reply:
OK, perfect.